### PR TITLE
Work-arounds for python 2.6 support

### PR DESCRIPTION
**Pull request by paalmbj**
This pull request fixes some compatibility problems with Python 2.6. 

I had problems that version_info.major is not defined, and that ElementTree.ParseError does not exist on 2.6.

### CODE REVIEW

**Comment by tcezard** [comment]
One could really argue you should upgrade to 2.7 but otherwise I don't have an issue with this

**Comment by paalmbj** [comment]
I agree that 2.6 is very old, but it's the default version on RHEL 6, which is the newest supported platform for the LIMS.

File: genologics/lims.py [hunk 1]
```diff
@@ -18,7 +18,7 @@
 # python 2.7, 3+ compatibility
 from sys import version_info
 
-if version_info.major == 2:
```
Comment (Galithil) [inline genologics/lims.py]: I kinda prefer `major` over `[0]` as it's clearer. Is there a specific reason to use [0] ?

Comment (paalmbj) [inline genologics/lims.py]: `major` was introduced in 2.7. (otherwise, I agree that it's clearer)

**Comment by Galithil** [comment]
Okay. All the PRs seem fine and pass the tests. 
We're going to need an update of the version.py, though. I'll update pypi afterwards.

**Comment by tcezard** [comment]
Do we want to add 2.6 in the travis.yaml to make sure all tests runs with python2.6 ?

**Comment by Galithil** [comment]
That would make sense, yes.

**Comment by paalmbj** [comment]
There are some `version_info.major` references in the test_descriptors.py and test_lims.py, but even changing those, I'm having trouble getting the tests to run. May be an old version of unittest, but I'll look into it.

**Comment by tcezard** [comment]
Yes same things here.
I don't think it is an issue with the tests though.
I think genologics has become genuinely incompatible with python 2.6

**Comment by tcezard** [comment]
I'm worried that you won't be able to use the current version of genologics with python2.6 at all.
For example: 
https://github.com/SciLifeLab/genologics/blob/master/genologics/lims.py#L549
```python
def write(self, outfile, etree):
        "Write the ElementTree contents as UTF-8 encoded XML to the open file."
        etree.write(outfile, encoding='utf-8', xml_declaration=True)
```
```xml_declaration``` does not exist in python2.6, it exist in python2.7 but is still compatible with python 2.6 behavior  if missing and it is mandatory in python3 if you want the xml declaration to be present.

**Comment by paalmbj** [comment]
Thanks for explaining the differences in `xml_declaration`. I opted to monkey-patch it in the top of the code, to keep all the compatibility stuff in one place. In our production system I have just removed xml_declaration. Unfortunately I missed that in my checks prior to the pull request (but then the tests caught it). I'll see if I can do something about the tests while keeping it reasonably clean, so don't merge it yet.

**Comment by tcezard** [comment]
The test coverage is very limited so there will be section of the code that might not work with python2.6 and you won't know until you try to run them locally.

We also run clarity on Centos6 but always install python2.7 and python3.4 from source and run all our EPPs with them.

**Comment by paalmbj** [comment]
Sorry that it's taking so long. It seems like most issues can be solved without making the testing code much uglier, but I don't know how to deal with

    assert isinstance(r, xml.etree.ElementTree.Element)

May be a while still until something happens.

**Comment by paalmbj** [comment]
I've modified some code for compatibility, not sure if you'll like it. Feel free to reject it if it's too ugly, but I may try to improve it.

**Comment by paalmbj** [comment]
By the way, still getting errors related to the EPP logger. Seems to be related to it not being installed as a package. Not sure if this is expected.

**Comment by Galithil** [comment]
I strongly recomment not using the EppLogger, it catches too many things : if you use pdb, for instance, it catches its output and makes it a pain to debug EPPs

**Comment by paalmbj** [comment]
Ok, good to know. I'm not using the EppLogger in production, but it's part of the tests, and those are failing in my local install. But they are also failing on Python 3 for me.

**Comment by Galithil** [comment]
Can I get an update on the status of this PR ?

**Comment by paalmbj** [comment]
Sorry for the late reply. I think this should be working. I had to change the way to determine whether `lims.get` returns a valid ElementTree.Element, because of compatibility. So that is the deal with changing

       assert isinstance(r, xml.etree.ElementTree.Element)

to

        assert r is not None
        assert callable(r.find)
        assert hasattr(r.attrib, '__getitem__') 

I'd argue that any object with a find() function and a attrib is probably good enough, for the purpose of this unit test.

So the status is that it seems to work, but the test is slightly altered.

**Comment by tcezard** [comment]
Hi @paalmbj,
I can see that it would work, however I really worried about having to jump through many hoops to support python 2.6 in the future.

FYI python 2.6 was released in 2008 and stopped being supported in 2013. I don't think we should encourage people to keep using it especially when it's quite straightforward to move to python 2.7 since it's backward compatible with 2.6 and is easy to install either from source or with a package manager.

I'm not saying there is anything wrong with this PR but it's not going in the right direction.

**Comment by Galithil** [comment]
I agree with you @tcezard, but @paalmbj raises an interesting point nonetheless. Since Clarity exists as a hosted solution, you might not _always_ have the possibility to update your python installation.
Maybe asking Genologics which version they support is the way to go.

**Comment by paalmbj** [comment]
I'm fine to let this one go, but I'm not convinced that updating to 2.7 is straightforward. @tcezard  Are you replacing the system python, or having a separate install and specifying that in all EPPs?

**Comment by tcezard** [comment]
I usually install python from source then create a virtual-env that I simlink into glsai's home/bin (which is in glsai's PATH).
If you don't want to change all you're EPP's command to python I would change glsai's PATH to make sure that the python you want to use is the first in the PATH.
You can of course use full path in you EPPs.

**Comment by Galithil** [comment]
We use a conda environment deployed in glsai's home, and since we make the
epps executable, you can either call their conda-env executable, or call
the right python as @tcezard says.

**Comment by paalmbj** [comment]
Ok, thanks for the explanations.

**Comment by tcezard** [comment]
An I assume you have ```#!/usr/bin/env python``` in your shebang  not ```#!/usr/bin/python```. 
The former take python from your PATH the second take python from the hard-coded path

**Comment by Galithil** [comment]
No, not always. What matters is how you call them.
It usually looks like that : "/home/glsai/anaconda/envs/<env_name>/bin/<epp_name>", since we use easy_install to mark epps as executable on installs.

**Comment by sstadick** [comment]
Hello! This might be a little outside the scope of this thread, but we've totally worked around 2.6 support by just not using it. All feedback is appreciated. 

The method that we've worked out to avoid being stuck with 2.6 / using the system python install has been to use the RHEL Software Collection python install: https://www.softwarecollections.org/en/scls/rhscl/python33/ (there's also python2.7)

We also use a virtualenv inside of that. So an epp call would look like: 
```
bash -c "scl enable python33 'source /opt/test/lims_venv/bin/activate; /opt/test/pQC/bin/alignment_qc.py --pluid={processLuid} --username={username} --password={password} --api_uri=******* 2> {compoundOutputFileLuid0}'"
```

This allows us to use python3 as well as any packages we feel like keeping track of.

**Comment by varun-ramraj** [comment]
As requested by @tcezard , I'm jumping into this thread! :)

I see myriad different setups, but most customers usually stick with 2.6. This, of course, causes a problem for me philosophically, because I prefer 2.7 (not least because argparse is a standard package). My job is to work with whatever you guys use, really, but I agree that 2.7 should be the standard version whenever possible. Therefore, my thoughts are as follows:

1) The custom 2.7 install in glsai's environment is probably the cleanest approach. I would caution against overwriting any python paths with the new location, however; instead, use the full paths when possible, as was suggested by many of you (or source a set of Python paths for your specific Python version using a wrapper bash script for the EPP, which then calls your Python code from within).

2) Python/Conda et c. doesn't matter much to me at all; from my point of view, the API scripts I help with are platform-agnostic in this sense. If the script runs, the Python/Conda is probably fine (naive, maybe, but fine for basic scripts).

3) Our glsapiutil libraries are written for Python 2.x, so in regards to Python 3, it is technically unsupported. One of my goals is to make the new glsapiutil libraries work with P3. In any case, if you're trying anything out with P3, definitely install it separately and don't use it in place of the system Python! I'm also happy to receive feedback on your experiences with P3 (silence in this case means P3 largely works, which is good feedback for me to take back to the devs) :)

4) I saw some comments above about easy_install. My choice here is definitely pip over easy_install whenever possible (especially when using system Python).

I'm aware that most of you on this thread are power users, so please keep the discussion/feedback coming, as it will help us make design decisions for future API-related scripts. Thanks!

**Comment by varun-ramraj** [comment]
Quick extra point here, for EPPs, I've taken to including the python invocation so it's easier to see which Python is being used (rather than marking the script as executable and having to look at the shebang to determine which Python is being invoked).

`bash -l -c "/path/to/python /path/to/pyscript.py -a arg1 -b -arg2 ..."`

**Comment by Galithil** [comment]
I guess it's time to get done with this.

I do not want to spend too much time supporting python 2.6. I will update the readme to reflect that.
This being said, the current build dies  horribly with python 2.6, and that's not great. I'm merging this in order to maintain the compatibility for now.

**Pull request merged**